test/integration/controller_binding_test.go

```diff
@@ -0,0 +1,413 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"reflect"
+	"testing"
+
+	// avoid error `servicecatalog/v1beta1 is not enabled`
+	_ "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/install"
+
+	osb "github.com/pmorie/go-open-service-broker-client/v2"
+	fakeosb "github.com/pmorie/go-open-service-broker-client/v2/fake"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1beta1"
+	"github.com/kubernetes-incubator/service-catalog/test/util"
+	"k8s.io/apimachinery/pkg/runtime"
+	"net/http"
+)
+
+// TestCreateServiceInstanceNonExistentClusterServiceClassOrPlan tests that a ServiceInstance gets
+// a Failed condition when the service class or service plan it references does not exist.
+func TestCreateServiceBinding(t *testing.T) {
+	cases := []struct {
+		name                string
+		instanceName        *string
+		instanceClassName   *string
+		instancePlanName    *string
+		instanceNotReady    bool
+		asyncForInstances   bool
+		asyncForBindings    bool
+		expectedErrorReason string
+		expectedFailure     bool
+		nonbindablePlan     bool
+		duplicateParameters bool
+	}{
+		{
+			name: "happy",
+		},
+		{
+			name:                "non-existent service instance name",
+			instanceName:        strPtr("nothereinstance"),
+			expectedErrorReason: "ReferencesNonexistentInstance",
+		},
+		{
+			name:                "invalid service instance name",
+			instanceName:        strPtr(""),
+			expectedErrorReason: "ReferencesNonexistentInstance",
+			expectedFailure:     true,
+		},
+		// TODO: this test is flaky. Which could be a bug?
```

Comment:
This is flaky because you are not leaving the ServiceInstance in an in-progress state. I presume that you are trying to set up the test so that the response to the poll request for the Provision always comes back with a state of InProgress. You have the broker reaction set up to only return InProgress if the OperationKey sent in the LastOperation request is `testInstanceLastOperation`. However, you are not including an OperationKey in the response to the Provision request. Consequently, no OperationKey is sent in the LastOperation request and the broker is responding with a state of Success.

It looks like you tried to set the OperationKey by adding it to the status of the instance in `getTestInstance`. However, the API Server ignores everything in the Status when creating a ServiceInstance.

Reply:
Ah. Ok I took another crack at it but then I got the service catalog controller in a bad state. Will look more tomorrow.